File: src/backend/routers/systems.py
"""
System deployment routes for the Trinity backend.

Provides endpoints for deploying multi-agent systems from YAML manifests.
"""
import logging

from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import JSONResponse, PlainTextResponse

from models import (
    BundledManifestDetail,
    BundledManifestSummary,
    User,
    SystemDeployRequest,
    SystemDeployResponse,
)
from database import db
from dependencies import get_current_user, require_role
from services.system_service import (
    system_member_names,
    deploy_manifest,
    list_bundled_manifests,
    read_bundled_manifest,
)
from services.docker_utils import container_stop

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/systems", tags=["systems"])


@router.post("/deploy", response_model=SystemDeployResponse)
async def deploy_system(
    body: SystemDeployRequest,
    request: Request,
    current_user: User = Depends(require_role("creator"))
):
    """
    Deploy a multi-agent system from YAML manifest.

    Thin HTTP wrapper over `services.system_service.deploy_manifest`
    (orchestration moved there in trinity-enterprise#124 so the first-run
    seeder can reuse it without importing a router — Invariant #1).

    Best-effort by default (trinity-enterprise#125): a per-agent create failure
    is reported in `failed[]` and the remaining agents still deploy; callers
    must check `status`, not just the HTTP code. `status` is "deployed" (all
    created), "partial" (some failed, HTTP 200), "failed" (none created,
    HTTP 500 with the full report as the body), "valid" (dry_run, no blockers)
    or "invalid" (dry_run, blockers in `failed` — #1841).

    Note that `status` describes AGENT CREATION only: a folder, permission,
    schedule, tag or start failure lands in `warnings[]` while `status` stays
    "deployed" (trinity-enterprise#125), so a caller rendering the outcome must
    surface `warnings` too.

    Args:
        body.manifest: YAML string defining the system
        body.dry_run: If true, validate only without creating agents
        body.strict: If true, abort on the first agent-create failure
            (legacy behavior), preserving the failure's original status code

    Returns:
        SystemDeployResponse with created agents, per-agent failures, and
        configuration summary
    """
    result = await deploy_manifest(
        body.manifest,
        current_user,
        request,
        dry_run=body.dry_run,
        strict=body.strict,
    )

    # Total failure: nothing was created — non-2xx so code-only callers
    # (curl -f) don't read it as success (trinity-enterprise#125).
    if result.status == "failed":
        return JSONResponse(status_code=500, content=result.model_dump())

    return result


@router.get("")
async def list_systems(current_user: User = Depends(get_current_user)):
    """
    List all systems (agents grouped by prefix).

    Groups agents by system prefix (before first '-').
    Returns system summaries with agent counts and details.
    """
    try:
        # Import here to avoid circular dependency
        from routers.agents import get_accessible_agents

        # Get all agents user can access
        agents = get_accessible_agents(current_user)

        # Group by system prefix
        systems_dict: dict = {}
        for agent in agents:
            # Extract system prefix (everything except last component after final '-')
            # Example: "my-system-abc-worker1" -> "my-system-abc"
            if '-' in agent['name']:
                parts = agent['name'].split('-')
                prefix = '-'.join(parts[:-1])  # All parts except the last (short name)
                if prefix not in systems_dict:
                    systems_dict[prefix] = {
                        "name": prefix,
                        "agents": [],
                        "agent_count": 0,
                        "created_at": agent.get('created_at')
                    }
                systems_dict[prefix]["agents"].append({
                    "name": agent['name'],
                    "status": agent.get('status', 'unknown'),
                    "template": agent.get('template')
                })
                systems_dict[prefix]["agent_count"] += 1

        # Sort by created_at (newest first)
        systems = list(systems_dict.values())
        systems.sort(key=lambda s: s.get('created_at') or '', reverse=True)

        return {"systems": systems}

    except Exception as e:
        logger.exception(f"Failed to list systems: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to list systems: {str(e)}")


# ============================================================================
# Bundled-manifest catalog (ent#126)
#
# ⚠️ Invariant #4 — BOTH routes must stay above the parameterized routes below,
# and for TWO separate collisions:
#   * `GET /manifests` would otherwise be captured by `GET /{system_name}` and
#     404 as "system 'manifests' not found" — a silent, plausible-looking failure.
#   * `GET /manifests/manifest` would otherwise be captured by
#     `GET /{system_name}/manifest` with system_name="manifests".
# Both collisions are covered by real request tests in
# tests/unit/test_ent126_manifest_catalog.py.
#
# Naming adjacency worth knowing: `/api/systems/manifests` (this bundled catalog)
# and `/api/systems/{name}/manifest` (export a DEPLOYED system as YAML) read
# alike and are unrelated.
#
# `require_role("creator")` mirrors POST /deploy rather than the looser
# get_current_user on the neighbouring list/get routes: a surface you cannot act
# on should not be advertised, and require_role also rejects connector principals.
# ============================================================================

@router.get("/manifests", response_model=list[BundledManifestSummary])
async def list_manifests(
    current_user: User = Depends(require_role("creator"))
):
    """
    List the system manifests bundled in `config/manifests/`.

    Read-only. Fail-soft per file: a manifest that cannot be parsed, validated or
    read is returned with `valid: false` and a short `reason` rather than failing
    the whole listing — one bad file must not hide the others.

    `valid: true` means parse + validate + the same side-effect-free
    template/resource preflight the dry-run uses all passed. It still cannot
    promise a `github:`-template manifest deploys, because those are not probed.
    """
    return list_bundled_manifests()


@router.get("/manifests/{manifest_id}", response_model=BundledManifestDetail)
async def get_manifest(
    manifest_id: str,
    current_user: User = Depends(require_role("creator"))
):
    """
    Read one bundled manifest's YAML, for loading into the install editor.

    A malformed id is 400; an unknown one is 404.
    """
    try:
        return read_bundled_manifest(manifest_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except FileNotFoundError:
        raise HTTPException(
            status_code=404, detail=f"Bundled manifest '{manifest_id}' not found"
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.exception(f"Failed to read bundled manifest '{manifest_id}': {e}")
        raise HTTPException(
            status_code=500, detail=f"Failed to read bundled manifest: {str(e)}"
        )


@router.get("/{system_name}")
async def get_system(
    system_name: str,
    current_user: User = Depends(get_current_user)
):
    """
    Get system details with all agents.

    Returns detailed information about a system including all its agents,
    permissions, folders, and schedules.
    """
    try:
        from routers.agents import get_accessible_agents

        # Get all agents user can access
        agents = get_accessible_agents(current_user)

        # Filter agents by system prefix
        # #2373: membership by TAG (deploy records it), prefix only as a
        # narrowed fallback. `startswith(f"{system_name}-")` also captured every
        # agent of a system named `acme-extra` — including on `restart`, which
        # stops and starts containers.
        _members = set(system_member_names(system_name, [a['name'] for a in agents]))
        system_agents = [agent for agent in agents if agent['name'] in _members]

        if not system_agents:
            raise HTTPException(
                status_code=404,
                detail=f"System '{system_name}' not found or no accessible agents"
            )

        # Get detailed info for each agent
        detailed_agents = []
        for agent in system_agents:
            agent_detail = {
                "name": agent['name'],
                "status": agent.get('status', 'unknown'),
                "template": agent.get('template'),
                "created_at": agent.get('created_at')
            }

            # Try to get additional details (permissions, folders, schedules)
            try:
                # Get permissions
                perms = db.get_agent_permissions(agent['name'])
                agent_detail["permissions"] = [p["target_agent"] for p in perms]

                # Get folders config
                folder_config = db.get_agent_folder_config(agent['name'])
                if folder_config:
                    agent_detail["folders"] = {
                        "expose": folder_config["expose_enabled"],
                        "consume": folder_config["consume_enabled"]
                    }

                # #2373: `db.get_agent_schedules` DOES NOT EXIST — the facade
                # exposes `list_agent_schedules`, and `database.py` deliberately
                # has no `__getattr__` fallback. The AttributeError was swallowed
                # by the `except Exception` below, so every response omitted
                # `schedules` for every agent and logged one warning each, while
                # `tests/test_systems.py` never asserted on the key. Exactly the
                # failure mode the db facade's own comment warns about.
                agent_detail["schedules"] = db.list_agent_schedules(agent['name'])

            except Exception as e:
                logger.warning(f"Failed to get details for agent {agent['name']}: {e}")

            detailed_agents.append(agent_detail)

        return {
            "name": system_name,
            "agent_count": len(detailed_agents),
            "agents": detailed_agents
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.exception(f"Failed to get system '{system_name}': {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get system: {str(e)}")


@router.post("/{system_name}/restart")
async def restart_system(
    system_name: str,
    request: Request,
    current_user: User = Depends(require_role("creator"))
):
    """
    Restart all agents in a system.

    Stops and starts every member. Useful after configuration changes.

    #2373: gated on `require_role("creator")`. It was bare `get_current_user` —
    below the gate on `POST /deploy` and below even the READ-ONLY
    bundled-catalog routes — so any authenticated principal, including
    `role: user`, could stop and start every container in any system whose
    agents it could see. A mutating fleet-wide verb under a lighter gate than
    the catalog it reads is an oversight, not a decision. `require_role` also
    rejects agent principals since #1890, which matters here because an
    agent-scoped MCP key resolves to its owner carrying the owner's role.
    """
    try:
        from routers.agents import get_accessible_agents, start_agent_internal
        from services.docker_service import get_agent_container

        # Get all agents user can access
        agents = get_accessible_agents(current_user)

        # Filter agents by system prefix
        # #2373: membership by TAG (deploy records it), prefix only as a
        # narrowed fallback. `startswith(f"{system_name}-")` also captured every
        # agent of a system named `acme-extra` — including on `restart`, which
        # stops and starts containers.
        _members = set(system_member_names(system_name, [a['name'] for a in agents]))
        system_agents = [agent for agent in agents if agent['name'] in _members]

        if not system_agents:
            raise HTTPException(
                status_code=404,
                detail=f"System '{system_name}' not found or no accessible agents"
            )

        restarted = []
        failed = []

        for agent in system_agents:
            agent_name = agent['name']
            try:
                # Stop agent
                if agent.get('status') == 'running':
                    container = get_agent_container(agent_name)
                    if container:
                        await container_stop(container)
                        logger.info(f"Stopped agent '{agent_name}' for system restart")

                # Start agent (with Trinity injection)
                await start_agent_internal(agent_name)
                restarted.append(agent_name)
                logger.info(f"Restarted agent '{agent_name}' for system '{system_name}'")

            except Exception as e:
                logger.error(f"Failed to restart agent '{agent_name}': {e}")
                failed.append(agent_name)

        return {
            "restarted": restarted,
            "failed": failed
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.exception(f"Failed to restart system '{system_name}': {e}")
        raise HTTPException(status_code=500, detail=f"Failed to restart system: {str(e)}")


@router.get("/{system_name}/manifest", response_class=PlainTextResponse)
async def get_system_manifest(
    system_name: str,
    current_user: User = Depends(get_current_user)
):
    """
    Export system as YAML manifest.

    Generates a YAML manifest from the current system configuration.
    Useful for backup, documentation, or replicating systems.
    """
    try:
        from routers.agents import get_accessible_agents
        from services.system_service import export_manifest

        # Get all agents user can access
        agents = get_accessible_agents(current_user)

        # Filter agents by system prefix
        # #2373: membership by TAG (deploy records it), prefix only as a
        # narrowed fallback. `startswith(f"{system_name}-")` also captured every
        # agent of a system named `acme-extra` — including on `restart`, which
        # stops and starts containers.
        _members = set(system_member_names(system_name, [a['name'] for a in agents]))
        system_agents = [agent for agent in agents if agent['name'] in _members]

        if not system_agents:
            raise HTTPException(
                status_code=404,
                detail=f"System '{system_name}' not found or no accessible agents"
            )

        # Export manifest
        yaml_content = export_manifest(system_name, system_agents)

        return yaml_content

    except HTTPException:
        raise
    except Exception as e:
        logger.exception(f"Failed to export manifest for system '{system_name}': {e}")
        raise HTTPException(status_code=500, detail=f"Failed to export manifest: {str(e)}")

File: src/mcp-server/src/tools/systems.ts
/**
 * System Management Tools
 *
 * MCP tools for deploying and managing multi-agent systems via YAML manifests.
 */

import { z } from "zod";
import { TrinityClient, ApiError } from "../client.js";
import type { McpAuthContext } from "../types.js";

/**
 * Collapse a teardown refusal into an honest, actionable shape (ent#454).
 *
 * `teardown_system` proxies an ENTITLEMENT-GATED route while knowing nothing
 * about entitlement, and the refusals it can meet are DIFFERENT operator
 * situations that must not be flattened into "it failed":
 *
 *   404  the route is absent — an OSS-only build with no teardown module
 *   403  mounted but not licensed for this instance
 *   403  an agent-scoped key called a human-only verb
 *   503  membership could not be verified; RETRY is the remedy
 *
 * It never throws: a thrown error reaches the agent as an opaque transport
 * failure it cannot reason about. The `{success: false, …flags}` shape is the
 * `call_a2a_agent` / `fetch_credential` precedent, where the flags ARE the
 * structure.
 */
function teardownFailure(error: unknown): string {
  const status = error instanceof ApiError ? error.status : undefined;
  const raw = error instanceof Error ? error.message : String(error);
  // `ApiError.body` is the raw response body, retained verbatim for exactly
  // this purpose (ent#443). Re-deriving it by regex off `message` is what
  // breaks the day the message format changes — its own docstring says so.
  const body = error instanceof ApiError ? error.body : undefined;
  const flags: Record<string, unknown> = {};
  let message = raw;

  // A `failed` teardown arrives as a 500 whose body IS the report, and that
  // report is the only actionable output — so it is a RESULT, not an error.
  if (status === 500 && body) {
    try {
      const report = JSON.parse(body);
      if (report && report.status === "failed") {
        return JSON.stringify(report, null, 2);
      }
    } catch {
      // Not JSON — fall through to the flag shape below.
    }
  }

  if (status === 404) {
    flags.not_available = true;
    message =
      "System teardown is not available on this platform. Remove members "
      + "individually with DELETE /api/agents/{name}.";
  } else if (status === 403) {
    flags.not_permitted = true;
    message =
      "Refused. System teardown is licensed separately AND is human-only: an "
      + "agent-scoped key cannot remove a fleet, because that is the bulk form "
      + "of an operation that is spawn-scoped one agent at a time. Ask a human "
      + "operator with the creator role.";
  } else if (status === 503) {
    flags.membership_unverified = true;
    flags.retryable = true;
    message =
      "Refused: system membership could not be verified, so teardown will not "
      + "derive a removal set from name prefixes alone. Nothing was removed. "
      + "Retry once the platform database is reachable.";
  } else if (status === 400) {
    // The confirmed set named no current member — re-preview and re-confirm.
    flags.bad_request = true;
  }

  return JSON.stringify({ success: false, error: message, detail: raw, ...flags }, null, 2);
}

/**
 * Create system management tools with the given client
 * @param client - Base Trinity client (provides base URL, no auth when requireApiKey=true)
 * @param requireApiKey - Whether API key authentication is enabled
 */
export function createSystemTools(
  client: TrinityClient,
  requireApiKey: boolean
) {
  /**
   * Get Trinity client with appropriate authentication
   * When requireApiKey is true, REQUIRES MCP API key from auth context
   * When requireApiKey is false, uses the base client (backward compatibility)
   */
  const getClient = (authContext?: McpAuthContext): TrinityClient => {
    if (requireApiKey) {
      // MCP API key is REQUIRED - no fallback
      if (!authContext?.mcpApiKey) {
        throw new Error("MCP API key authentication required but no API key found in request context");
      }
      // Create new client instance authenticated with user's MCP API key
      const userClient = new TrinityClient(client.getBaseUrl());
      userClient.setToken(authContext.mcpApiKey);
      return userClient;
    }
    // API key auth disabled - use base client (backward compatibility)
    return client;
  };

  return {
    // ========================================================================
    // deploy_system - Deploy multi-agent system from YAML manifest
    // ========================================================================
    deploySystem: {
      name: "deploy_system",
      description:
        "Deploy a multi-agent system from a YAML manifest. " +
        "The manifest defines the system name, agents, permissions, schedules, and shared folders. " +
        "Supports dry_run mode for validation without deployment — a dry run resolves each `local:` template, "
        + "so `status: 'invalid'` with a populated `failed[]` means the manifest would not deploy cleanly. " +
        "Agents are created with naming pattern '{system}-{agent}' (e.g., 'content-production-orchestrator'). " +
        "Supports permission presets: 'full-mesh', 'orchestrator-workers', 'none', or explicit rules. " +
        "Deploy is best-effort: check `status` ('deployed' | 'partial' | 'failed') and `failed[]` " +
        "in the response — a partial deploy still creates the remaining agents. " +
        "Pass strict: true to restore abort-on-first-error.",
      parameters: z.object({
        manifest: z
          .string()
          .describe(
            "YAML manifest as a string. Format:\n" +
            "name: system-name\n" +
            "description: System description\n" +
            "prompt: System-wide instructions (optional)\n" +
            "agents:\n" +
            "  orchestrator:\n" +
            "    template: github:Org/repo\n" +
            "    folders: {expose: true, consume: true}\n" +
            "    schedules: [{name: daily, cron: '0 9 * * *', message: '...'}]\n" +
            "permissions:\n" +
            "  preset: full-mesh  # or orchestrator-workers, none, explicit"
          ),
        dry_run: z
          .boolean()
          .optional()
          .describe(
            "If true, validates the manifest without creating agents. " +
            "Returns preview of agents to be created and warnings."
          ),
        strict: z
          .boolean()
          .optional()
          .describe(
            "If true, abort the deploy on the first agent-create failure " +
            "(legacy behavior) instead of the default best-effort " +
            "continue-and-report."
          ),
      }),
      execute: async (
        { manifest, dry_run, strict }: { manifest: string; dry_run?: boolean; strict?: boolean },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        // Call backend deploy endpoint
        const response = await apiClient.request<{
          status: string; // "deployed" | "partial" | "failed" | "valid" (dry_run) | "invalid" (dry_run, #1841)
          system_name: string;
          agents_created: string[];
          agents_to_create?: Array<{ name: string; short_name: string; template: string }>;
          prompt_updated: boolean;
          permissions_configured?: number;
          schedules_created?: number;
          warnings: string[];
          failed?: Array<{
            name: string;
            short_name: string;
            template: string;
            reason: string;
            status_code?: number;
          }>;
        }>("POST", "/api/systems/deploy", {
          manifest,
          dry_run: dry_run || false,
          strict: strict || false,
        });

        return JSON.stringify(response, null, 2);
      },
    },

    // ========================================================================
    // list_systems - List all deployed systems (agents grouped by prefix)
    // ========================================================================
    listSystems: {
      name: "list_systems",
      description:
        "List all deployed systems with their agents. " +
        "Groups agents by system prefix (before first '-'). " +
        "For example, agents 'content-production-orchestrator' and 'content-production-writer' " +
        "are grouped under system 'content-production'. " +
        "Returns system summaries with agent counts and details.",
      parameters: z.object({}),
      execute: async (_params: unknown, context?: { session?: McpAuthContext }) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const response = await apiClient.request<{
          systems: Array<{
            name: string;
            agent_count: number;
            agents: Array<{
              name: string;
              status: string;
              template?: string;
            }>;
            created_at?: string;
          }>;
        }>("GET", "/api/systems");

        return JSON.stringify(response, null, 2);
      },
    },

    // ========================================================================
    // restart_system - Restart all agents in a system
    // ========================================================================
    restartSystem: {
      name: "restart_system",
      description:
        "Restart all agents belonging to a system. " +
        "Membership is resolved from the system's tag when the agents carry one, " +
        "falling back to the '{system_name}-*' name prefix. " +
        "Stops then starts each member. " +
        "Useful after configuration changes (credentials, schedules, shared folders). " +
        "Returns list of successfully restarted agents and any failures. " +
        "Requires the 'creator' role AND a human caller: an agent-scoped key is " +
        "refused with 403 (#2373), because this restarts every member without the " +
        "per-agent spawn-scope check that start_agent/stop_agent apply one at a time.",
      parameters: z.object({
        system_name: z
          .string()
          .describe(
            "System to restart (e.g., 'content-production'). " +
            "Members are the agents tagged with the system, else those matching " +
            "'{system_name}-*'."
          ),
      }),
      execute: async (
        { system_name }: { system_name: string },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const response = await apiClient.request<{
          restarted: string[];
          failed: string[];
        }>("POST", `/api/systems/${encodeURIComponent(system_name)}/restart`);

        return JSON.stringify(response, null, 2);
      },
    },

    // ========================================================================
    // get_system_manifest - Export system configuration as YAML manifest
    // ========================================================================
    getSystemManifest: {
      name: "get_system_manifest",
      description:
        "Generate a YAML manifest for a deployed system. " +
        "Reconstructs the system configuration from current agent settings. " +
        "Useful for backup, documentation, or replicating systems. " +
        "Returns YAML string that can be used with deploy_system.",
      parameters: z.object({
        system_name: z
          .string()
          .describe(
            "System prefix to export (e.g., 'content-production'). " +
            "All agents matching '{system_name}-*' will be included."
          ),
      }),
      execute: async (
        { system_name }: { system_name: string },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        // Backend returns YAML as text/plain
        const yaml = await apiClient.request<string>(
          "GET",
          `/api/systems/${encodeURIComponent(system_name)}/manifest`
        );

        return yaml;
      },
    },

    // ========================================================================
    // teardown_system - Remove a deployed system (ent#454)
    //
    // License-blind proxy over the entitlement-gated
    // DELETE /api/enterprise/system-teardown/{name}. This module knows nothing
    // about entitlement; it reports what the route says (see teardownFailure).
    // ========================================================================
    teardownSystem: {
      name: "teardown_system",
      description:
        "Remove a deployed multi-agent system: every member agent, plus the " +
        "system view its deploy auto-created. The inverse of deploy_system. " +
        "DESTRUCTIVE — dry_run defaults to TRUE, unlike deploy_system: call it " +
        "first, show the human the removal set, and pass the returned member " +
        "names back as `agents` to execute. " +
        "Each member goes through the standard per-agent soft-delete, so its " +
        "data, history and name are retained for the platform's recovery window " +
        "(180 days by default) and an admin can restore the record; recovery is " +
        "metadata-only, and members reported 'discarded' were ephemeral and have " +
        "NO recovery window. " +
        "Membership comes from the system's deploy tag, falling back to the " +
        "'{system_name}-*' name prefix — check each member's `evidence` field: " +
        "'prefix' means it matched by NAME ONLY and may belong to a different " +
        "system whose name starts the same way, so a human should confirm it. " +
        "Best-effort and honest: switch on `status` ('preview' | 'torn_down' | " +
        "'partial' | 'failed'), never the HTTP code, and read every member's " +
        "`outcome` — 'skipped' is not 'failed'. " +
        "Requires the 'creator' role AND a HUMAN caller: an agent-scoped key is " +
        "refused, because this removes N agents in one call without the " +
        "per-agent spawn-scope check that delete_agent applies one at a time. " +
        "Licensed separately — on a build without it, the tool says so rather " +
        "than failing opaquely.",
      parameters: z.object({
        system_name: z
          .string()
          .describe(
            "System to remove (e.g., 'content-production'). Members are the "
            + "agents tagged with the system, else those matching "
            + "'{system_name}-*'. Do NOT take this name from list_systems, "
            + "which groups by the last hyphen and can report a name that is "
            + "not a real system."
          ),
        dry_run: z
          .boolean()
          .optional()
          .describe(
            "Defaults to TRUE. Returns the complete removal set — members with "
            + "their current status, evidence and ephemeral flag, the system "
            + "view, the tag — and writes NOTHING. Pass false only to execute a "
            + "removal a human has confirmed."
          ),
        agents: z
          .array(z.string())
          .optional()
          .describe(
            "The member names confirmed for removal. Only the intersection with "
            + "current membership is removed; a name that is no longer a member "
            + "comes back 'skipped: not_a_member' and a current member you omit "
            + "comes back 'skipped: not_confirmed'. Omit to remove every current "
            + "member."
          ),
        strict: z
          .boolean()
          .optional()
          .describe(
            "Stop at the first member that fails, marking the remainder "
            + "'aborted'. Default is best-effort: one failure never aborts the "
            + "rest."
          ),
      }),
      execute: async (
        {
          system_name,
          dry_run,
          agents,
          strict,
        }: {
          system_name: string;
          dry_run?: boolean;
          agents?: string[];
          strict?: boolean;
        },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        // Preview-first unless the caller explicitly says otherwise. The
        // asymmetry with deploy_system (which defaults false) is deliberate:
        // the cost of an unwanted preview is a wasted round trip, and the cost
        // of an unwanted execute is a deleted fleet.
        const preview = dry_run === undefined ? true : dry_run;

        try {
          const response = await apiClient.request<unknown>(
            "DELETE",
            `/api/enterprise/system-teardown/${encodeURIComponent(system_name)}`
              + `?dry_run=${preview}`,
            // A DELETE body: `client.request` carries one, and the confirmed
            // set belongs here rather than in the query string because it is
            // the time-of-check/time-of-use seam the server re-validates.
            { agents: agents ?? null, strict: strict ?? false }
          );
          return JSON.stringify(response, null, 2);
        } catch (error) {
          return teardownFailure(error);
        }
      },
    },
  };
}
